### DISCUSSION TITLE
GPIO pin staying High after crash

### DISCUSSION THREAD

**Discussion by DKB234** [micropython/micropython#9515] [Hardware & Peripherals]
Using a ESP32 devKit CWRover to drive dc motor through a opto isolated relay, on a sun tracking solar array.
When the chip crashes it will leave the pin driving the Dc motor high, until I reboot it.
Have set up a WatchDog timer to catch faults and it works most of the time, but has failed twice now with not good out comes for the motor.
Is there a setting on a particular pin I can use to stop the Pin staying high after the crash.
I'm quite new to Python and ESP32 so am hoping it is just a rookie mistake.
I would appreciate any help
Thanks

**Comment by peterhinch** [comment]
I would want to know the answer to two questions:
 * Why is it crashing?
 * Why is the WDT failing to reset the system?

The fact that these are happening suggests there is something wrong with your hardware, possibly a power supply problem. The ESP32 can take quite a lot of current when using the WiFi. With a good PSU the ESP32 is capable of long term operation.

If you really can't stop it crashing you may need a hardware WDT that forces a reset. Or alternatively a hardware retriggerable monostable between the pin and the SSR so that the motor only runs if the pin is periodically pulsed. This is the brute force way to protect the motor.

**Reply by DKB234** [reply]
Thanks for your response 
I power the system from a  12v battery with 5days reserve solar charged.
Today's crash maybe from a problem logging in to WiFi MQTT server hoping I have fixed that issue

**Comment by robert-hh** [comment]
When the device crashes, what is it's state. Does it go though a hard reset? If yes, the pin should be set to input mode. Which Pin do you use for the motor? Some pins have an internal pull-up enabled after reset.

**Reply by DKB234** [reply]
When the WatchDog catches it yes it resets and pin goes low. I believe that the WiFi login problem failed before the pin was reset. I'm using pin 23 and pin22 to run the motor

**Comment by peterhinch** [comment]
What MQTT library are you using? The official ones are not very reliable in long-term running. Consider using [mqtt_as](https://github.com/peterhinch/micropython-mqtt/blob/master/mqtt_as/README.md) which specifically addresses issues of long term running.

**Reply by DKB234** [reply]
Didn't know of that one I've been using mqtt_robust with Mosquitto server on a Pi4
I just had a quick look at mqtt_as I will give it a try, Wih my limited knowledge of python and ESP32 it will take me some time to implement.
Thanks
